File: app/src/main/java/com/example/botondepanicov1/bluetooth/BuscandoDispositivosBluetooth.java
package com.example.botondepanicov1.bluetooth;

import android.Manifest;
import android.annotation.SuppressLint;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothManager;
import android.bluetooth.le.AdvertiseCallback;
import android.bluetooth.le.AdvertiseSettings;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.RemoteException;
import android.preference.PreferenceManager;
import android.util.Log;
import android.widget.ListView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import com.example.botondepanicov1.R;
import com.example.botondepanicov1.models.BluetoothFrame;
import com.example.botondepanicov1.wifi_direct.Ingredient;

import org.altbeacon.beacon.Beacon;
import org.altbeacon.beacon.BeaconConsumer;
import org.altbeacon.beacon.BeaconManager;
import org.altbeacon.beacon.BeaconParser;
import org.altbeacon.beacon.BeaconTransmitter;
import org.altbeacon.beacon.Identifier;
import org.altbeacon.beacon.MonitorNotifier;
import org.altbeacon.beacon.RangeNotifier;
import org.altbeacon.beacon.Region;

import java.net.NetworkInterface;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

public class BuscandoDispositivosBluetooth extends AppCompatActivity implements BeaconConsumer {
    // variables para la configuracion de los beacons
    private Beacon beacon;
    private BluetoothAdapter btAdapter;
    private BeaconManager beaconManager;
    private BeaconTransmitter beaconTransmitter;
    private static final String TAG = "Sergio";

    private ListView listaBluetooth;
    private List<BluetoothFrame> mLista = new ArrayList<>();
    private AdapterBluetooth mAdapter;

    //Direccion MAC
    String mac;
    String KEY_MAC = "MAC";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_buscando_dispositivos_bluetooth);
        //LEER PREFERENCIAS DE LA MAC
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(this);
        String datos = pref.getString(this.KEY_MAC, "No hay datos");
        Log.v("Sergio","MACKEY"+datos);
        // valida que se pueda optener la MAC, si no es posible genera el numero aleatorio
        if (datos.equals("No hay datos")){
            mac = trasformarMac();
            guardarMacAleatoria(mac);
        }{
            mac = datos;
        }

        //llama a funcion de validar permisos de localozacion
        checkPermission();
        listaBluetooth = findViewById(R.id.listBluetooth);
        btAdapter = BluetoothAdapter.getDefaultAdapter();
        beaconManager = BeaconManager.getInstanceForApplication(this);

        encenderBluetooth();
        //llama la funcion para la configuracion del beacon
        setupBeacon();
        //llama la funcion para envio de los beacons
        envio();

        beaconManager.getBeaconParsers().add(new BeaconParser()
                .setBeaconLayout("m:2-3=beac,i:4-19,i:20-21,i:22-23,p:24-24,d:25-25"));

        inicioDescubrimiento();
    }
    //transforma la MAC para que quede Hexadecimal
    public static String trasformarMac(){
        String mac = "";
        Log.v("Sergio", "uuid|" + obtenerMac().equals("")+"|");
        if(obtenerMac().equals("")){
            return alternativaMac();
        }
        else {
            String[] macArray = obtenerMac().split(":");
            for (int i = 0; i < macArray.length; i++) {
                if (macArray[i].length() == 1) {
                    macArray[i] = macArray[i] + "0";
                }
            }

            for (String i : macArray) {
                mac = mac + i;
            }
            return mac;
        }
    }
    // optiene la MAC de los dispositivos si es posible
    public static String obtenerMac() {
        try {
            List<NetworkInterface> all = Collections.list(NetworkInterface.getNetworkInterfaces());
            for (NetworkInterface nif : all) {
                if (!nif.getName().equalsIgnoreCase("wlan0")) continue;

                byte[] macBytes = nif.getHardwareAddress();
                if (macBytes == null) {
                    return "";
                }

                StringBuilder res1 = new StringBuilder();
                for (byte b : macBytes) {
                    res1.append(Integer.toHexString(b & 0xFF) + ":");
                }

                if (res1.length() > 0) {
                    res1.deleteCharAt(res1.length() - 1);
                }
                return res1.toString();
            }
        } catch (Exception ex) {
            Log.e("Error", ex.getMessage());
        }
        return "";
    }

    //si no es posible obtener la MAC guarda la aleatoria en preferencias
    public void guardarMacAleatoria(String mac){
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(this.KEY_MAC, mac);
        editor.apply();
    }

    // concatena los numeros aleatorios
    public static String alternativaMac(){
        return numeroAleatorio()+numeroAleatorio();
    }

    // se genera numero aletario de 6 digitos
    public static String numeroAleatorio(){
        long min_val = 100000;
        int max_val = 999999;
        double randomNum = Math.random() * ( max_val - min_val );
        System.out.println("Random Number: "+randomNum);
        return String.valueOf( (int) randomNum);
    }

    //termina la actividad cuando se hace uso de boton atras
    @Override
    public void onBackPressed() {
        beaconManager.unbind(BuscandoDispositivosBluetooth.this);
        this.finish();
    }

    //encender o apagar bluetooth
    private void encenderBluetooth(){
        if(!btAdapter.isEnabled()){
            btAdapter.enable();
        }
    }

    // inica el descubrimiento de dispositivos
    private void inicioDescubrimiento(){
        beaconManager.bind(BuscandoDispositivosBluetooth.this);
    }

    //valida que el bluetooth habilitado y compatible con el dispositivo
    private boolean isBluetoothLEAvailable() {
        return btAdapter != null && this.getPackageManager().hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE);
    }

    //valida que el bluetooth este encendido
    private boolean getBlueToothOn() {
        return btAdapter != null && btAdapter.isEnabled();
    }

    ///llama la funcion que valida si el bluetooth este encendido
    private void envio() {
        if (getBlueToothOn()) {
            Log.i(TAG, "isBlueToothOn");
            transmitIBeacon();
        } else if (!isBluetoothLEAvailable()) {
            Toast toast = Toast.makeText(this, "Bluetooth no disponible en su dispositivo.",Toast.LENGTH_LONG);
            toast.show();
        } else {
            Log.i(TAG, "BlueTooth is off");
            Toast toast = Toast.makeText(this, "Habilite bluetooth antes de transmitir iBeacon.",Toast.LENGTH_LONG);
            toast.show();
        }
    }

    // envio de beacons
    private void transmitIBeacon() {
        boolean isSupported;
        isSupported = btAdapter.isMultipleAdvertisementSupported();
        if (isSupported) {

            Log.v(TAG, "is support advertistment");
            if (beaconTransmitter.isStarted()) {
                beaconTransmitter.stopAdvertising();

            } else {
                beaconTransmitter.startAdvertising(beacon, new AdvertiseCallback() {

                    @Override
                    public void onStartFailure(int errorCode) {
                        Log.e(TAG, "Advertisement start failed with code: " + errorCode);
                    }

                    @Override
                    public void onStartSuccess(AdvertiseSettings settingsInEffect) {
                        Log.i(TAG, "Advertisement start succeeded." + settingsInEffect.toString());
                    }
                });

            }
        } else {
            Toast toast = Toast.makeText(this, "Su dispositivo no es compatible con LE Bluetooth.",Toast.LENGTH_LONG);
            toast.show();
        }
    }

    //configuracion del beacon concatenando la MAC
    private void setupBeacon() {
        String uuid = "954e6dac-5612-4642-b2d1-"+mac;
        Log.v("Sergio","uuid "+uuid);
        beacon = new Beacon.Builder()
                .setId1(uuid) // UUID for beacon
                .setId2("5") // Major for beacon
                .setId3("12") // Minor for beacon
                .setManufacturer(0x004C) // Radius Networks.0x0118  Change this for other beacon layouts//0x004C for iPhone
                .setTxPower(-56) // Power in dB
                .setDataFields(Arrays.asList(2L, 3L)) // Remove this for beacon layouts without d: fields
                .build();

        BluetoothManager btManager = (BluetoothManager) this.getSystemService(Context.BLUETOOTH_SERVICE);
        btAdapter = btManager.getAdapter ();

        beaconTransmitter = new BeaconTransmitter(this, new BeaconParser()
                .setBeaconLayout ("m:2-3=beac,i:4-19,i:20-21,i:22-23,p:24-24,d:25-25"));
    }

    //recepcion de beacons
    @Override
    public void onBeaconServiceConnect() {
        final Region region = new Region("myBeaons", Identifier.parse("2f234454-cf6d-4a0f-adf2-f4911ba9ffa6"), null, null);
        final Region region2 = new Region("myBeaons", null, null, null);

        beaconManager.setMonitorNotifier(new MonitorNotifier() {
            @Override
            public void didEnterRegion(Region region) {
                try {
                    Log.d(TAG, "didEnterRegion");
                    beaconManager.startRangingBeaconsInRegion(region);
                    beaconManager.startRangingBeaconsInRegion(region2);
                } catch (RemoteException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void didExitRegion(Region region) {
                try {
                    Log.d(TAG, "didExitRegion");
                    beaconManager.stopRangingBeaconsInRegion(region);
                    beaconManager.stopRangingBeaconsInRegion(region2);
                } catch (RemoteException e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void didDetermineStateForRegion(int i, Region region) {

            }
        });

        //se obtienen los datos recibidos de los beacons
        beaconManager.setRangeNotifier((beacons, region1) -> {
            for (Beacon oneBeacon : beacons) {

                Log.d(TAG, "distance: " + oneBeacon.getDistance() + " adrres:" + oneBeacon.getBluetoothAddress()
                        + " id:" + oneBeacon.getId1() + "/" + oneBeacon.getId2() + "/" + oneBeacon.getId3());

                mLista = eliminarDuplicados(mLista,oneBeacon);
                mAdapter = new AdapterBluetooth(BuscandoDispositivosBluetooth.this, R.layout.adapter_dispositivos_encontrados_wifi,mLista);
                listaBluetooth.setAdapter(mAdapter);
            }
        });

        try {
            beaconManager.startMonitoringBeaconsInRegion(region);
            beaconManager.startMonitoringBeaconsInRegion(region2);
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    // elimina el duplicados de dispositivos encontrados
    public List<BluetoothFrame> eliminarDuplicados(List<BluetoothFrame> lista, Beacon oneBeacon){
        Calendar c = Calendar.getInstance();
        @SuppressLint("SimpleDateFormat") SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        String strDate = sdf.format(c.getTime());
        for(int i = 0; i<mLista.size();i++){
            if(mLista.get(i).getIdentifier().equals(oneBeacon.getId1())){
                mLista.remove(i);
            }
        }

        BluetoothFrame frame = new BluetoothFrame();
        frame.setIdentifier(oneBeacon.getId1());
        frame.setDistance(oneBeacon.getDistance());
        frame.setDate(strDate);

        mLista.add(frame);
        return lista;
    }
    // valida los permisos de localizacion
    public void checkPermission() {
        if (Build.VERSION.SDK_INT >= 23) {
            if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED && checkSelfPermission(android.Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {

            } else {
                ActivityCompat.requestPermissions(this, new String[]{
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION,}, 1);
            }
        }
    }
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED && grantResults[1] == PackageManager.PERMISSION_GRANTED) {
        } else {
            checkPermission();
        }
    }
}

File: app/src/main/java/com/example/botondepanicov1/bluetooth/BuscandoDispositivosBluetooth.kt
package com.example.botondepanicov1.bluetooth

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseSettings
import android.content.pm.PackageManager
import android.location.LocationManager
import android.os.Build
import android.os.Bundle
import android.os.RemoteException
import android.preference.PreferenceManager
import android.util.Log
import android.widget.ExpandableListAdapter
import android.widget.ExpandableListView
import android.widget.ListView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.example.botondepanicov1.R
import com.example.botondepanicov1.adapters.IngredientAdapter
import com.example.botondepanicov1.core.Role
import com.example.botondepanicov1.models.BluetoothFrame
import com.example.botondepanicov1.wifi_direct.Ingredient
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import org.altbeacon.beacon.*
import java.net.NetworkInterface
import java.text.SimpleDateFormat
import java.util.*
import kotlin.experimental.and

class BuscandoDispositivosBluetooth : AppCompatActivity(), OnMapReadyCallback, BeaconConsumer {
    // variables para la configuracion de los beacons
    private var beacon: Beacon? = null
    private var btAdapter: BluetoothAdapter? = null
    private var beaconManager: BeaconManager? = null
    private var beaconTransmitter: BeaconTransmitter? = null
    private var listaBluetooth: ListView? = null
    private var mLista: MutableList<BluetoothFrame> = ArrayList()
    private var mAdapter: AdapterBluetooth? = null
    private lateinit var googleMap: GoogleMap

    //Direccion MAC
    var mac: String? = null
    var KEY_MAC = "MAC"
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_buscando_dispositivos_bluetooth)

        val childList = listOf(
            Ingredient().apply {
                username = "Homero Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "Lisa Sompson"
                distance = 123.1
                role = Role.RESCUER.ordinal
            },
            Ingredient().apply {
                username = "Bort Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "Marge Sompson"
                distance = 123.1
                role = Role.RESCUER.ordinal
            },
            Ingredient().apply {
                username = "Abraham Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "The Game Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "Homero Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "Lisa Sompson"
                distance = 123.1
                role = Role.RESCUER.ordinal
            },
            Ingredient().apply {
                username = "Bort Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "Marge Sompson"
                distance = 123.1
                role = Role.RESCUER.ordinal
            },
            Ingredient().apply {
                username = "Abraham Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "The Game Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            }, Ingredient().apply {
                username = "Homero Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "Lisa Sompson"
                distance = 123.1
                role = Role.RESCUER.ordinal
            },
            Ingredient().apply {
                username = "Bort Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "Marge Sompson"
                distance = 123.1
                role = Role.RESCUER.ordinal
            },
            Ingredient().apply {
                username = "Abraham Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            },
            Ingredient().apply {
                username = "The Game Sompson"
                distance = 123.1
                role = Role.SURVIVOR.ordinal
            }
        )
        val survivors = childList.filter { x -> x.role == Role.SURVIVOR.ordinal }
        val rescuers = childList.filter { x -> x.role == Role.RESCUER.ordinal }
        val collection = hashMapOf(Role.SURVIVOR to survivors, Role.RESCUER to rescuers)

        val expandableListView = findViewById<ExpandableListView>(R.id.expandable_list)
        val expandableListAdapter: ExpandableListAdapter = IngredientAdapter(this, collection, listOf(Role.SURVIVOR, Role.RESCUER))
        expandableListView.setAdapter(expandableListAdapter)

        expandableListView.setOnChildClickListener { _, _, group, child, _ ->
            val selected = expandableListAdapter.getChild(group, child)
            Toast.makeText(this, selected.toString(), Toast.LENGTH_LONG).show()
            true
        }

        locationUpdates()

        //LEER PREFERENCIAS DE LA MAC
        val pref = PreferenceManager.getDefaultSharedPreferences(this)
        val datos = pref.getString(KEY_MAC, "No hay datos")
        Log.v("Sergio", "MACKEY$datos")
        // valida que se pueda optener la MAC, si no es posible genera el numero aleatorio
        if (datos == "No hay datos") {
            mac = trasformarMac()
            guardarMacAleatoria(mac)
        }
        run { mac = datos }

        //llama a funcion de validar permisos de localozacion
        checkPermission()
        //TODO listaBluetooth = findViewById(R.id.listBluetooth)
        btAdapter = BluetoothAdapter.getDefaultAdapter()
        beaconManager = BeaconManager.getInstanceForApplication(this)
        encenderBluetooth()
        //llama la funcion para la configuracion del beacon
        setupBeacon()
        //llama la funcion para envio de los beacons
        envio()
        beaconManager!!.beaconParsers.add(
            BeaconParser()
                .setBeaconLayout("m:2-3=beac,i:4-19,i:20-21,i:22-23,p:24-24,d:25-25")
        )
        inicioDescubrimiento()
    }

    //si no es posible obtener la MAC guarda la aleatoria en preferencias
    fun guardarMacAleatoria(mac: String?) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        val editor = prefs.edit()
        editor.putString(KEY_MAC, mac)
        editor.apply()
    }

    //termina la actividad cuando se hace uso de boton atras
    override fun onBackPressed() {
        beaconManager!!.unbind(this@BuscandoDispositivosBluetooth)
        finish()
    }

    //encender o apagar bluetooth
    private fun encenderBluetooth() {
        if (!btAdapter!!.isEnabled) {
            btAdapter!!.enable()
        }
    }

    // inica el descubrimiento de dispositivos
    private fun inicioDescubrimiento() {
        beaconManager!!.bind(this@BuscandoDispositivosBluetooth)
    }

    //valida que el bluetooth habilitado y compatible con el dispositivo
    private val isBluetoothLEAvailable: Boolean
        private get() = btAdapter != null && this.packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE)

    //valida que el bluetooth este encendido
    private val blueToothOn: Boolean
        private get() = btAdapter != null && btAdapter!!.isEnabled

    ///llama la funcion que valida si el bluetooth este encendido
    private fun envio() {
        if (blueToothOn) {
            Log.i(TAG, "isBlueToothOn")
            transmitIBeacon()
        } else if (!isBluetoothLEAvailable) {
            val toast = Toast.makeText(
                this,
                "Bluetooth no disponible en su dispositivo.",
                Toast.LENGTH_LONG
            )
            toast.show()
        } else {
            Log.i(TAG, "BlueTooth is off")
            val toast = Toast.makeText(
                this,
                "Habilite bluetooth antes de transmitir iBeacon.",
                Toast.LENGTH_LONG
            )
            toast.show()
        }
    }

    // envio de beacons
    private fun transmitIBeacon() {
        val isSupported: Boolean
        isSupported = btAdapter!!.isMultipleAdvertisementSupported
        if (isSupported) {
            Log.v(TAG, "is support advertistment")
            if (beaconTransmitter!!.isStarted) {
                beaconTransmitter!!.stopAdvertising()
            } else {
                beaconTransmitter!!.startAdvertising(beacon, object : AdvertiseCallback() {
                    override fun onStartFailure(errorCode: Int) {
                        Log.e(TAG, "Advertisement start failed with code: $errorCode")
                    }

                    override fun onStartSuccess(settingsInEffect: AdvertiseSettings) {
                        Log.i(TAG, "Advertisement start succeeded.$settingsInEffect")
                    }
                })
            }
        } else {
            val toast = Toast.makeText(
                this,
                "Su dispositivo no es compatible con LE Bluetooth.",
                Toast.LENGTH_LONG
            )
            toast.show()
        }
    }

    //configuracion del beacon concatenando la MAC
    private fun setupBeacon() {
        val uuid = "954e6dac-5612-4642-b2d1-$mac"
        Log.v("Sergio", "uuid $uuid")
        beacon = Beacon.Builder()
            .setId1(uuid) // UUID for beacon
            .setId2("5") // Major for beacon
            .setId3("12") // Minor for beacon
            .setManufacturer(0x004C) // Radius Networks.0x0118  Change this for other beacon layouts//0x004C for iPhone
            .setTxPower(-56) // Power in dB
            .setDataFields(
                Arrays.asList(
                    2L,
                    3L
                )
            ) // Remove this for beacon layouts without d: fields
            .build()
        val btManager = this.getSystemService(BLUETOOTH_SERVICE) as BluetoothManager
        btAdapter = btManager.adapter
        beaconTransmitter = BeaconTransmitter(
            this, BeaconParser()
                .setBeaconLayout("m:2-3=beac,i:4-19,i:20-21,i:22-23,p:24-24,d:25-25")
        )
    }

    //recepcion de beacons
    override fun onBeaconServiceConnect() {
        val region =
            Region("myBeaons", Identifier.parse("2f234454-cf6d-4a0f-adf2-f4911ba9ffa6"), null, null)
        val region2 = Region("myBeaons", null, null, null)
        beaconManager!!.setMonitorNotifier(object : MonitorNotifier {
            override fun didEnterRegion(region: Region) {
                try {
                    Log.d(TAG, "didEnterRegion")
                    beaconManager!!.startRangingBeaconsInRegion(region)
                    beaconManager!!.startRangingBeaconsInRegion(region2)
                } catch (e: RemoteException) {
                    e.printStackTrace()
                }
            }

            override fun didExitRegion(region: Region) {
                try {
                    Log.d(TAG, "didExitRegion")
                    beaconManager!!.stopRangingBeaconsInRegion(region)
                    beaconManager!!.stopRangingBeaconsInRegion(region2)
                } catch (e: RemoteException) {
                    e.printStackTrace()
                }
            }

            override fun didDetermineStateForRegion(i: Int, region: Region) {}
        })

        //se obtienen los datos recibidos de los beacons
        beaconManager!!.setRangeNotifier { beacons: Collection<Beacon>, region1: Region? ->
            for (oneBeacon in beacons) {
                Log.d(
                    TAG, "distance: " + oneBeacon.distance + " adrres:" + oneBeacon.bluetoothAddress
                            + " id:" + oneBeacon.id1 + "/" + oneBeacon.id2 + "/" + oneBeacon.id3
                )

                mLista = eliminarDuplicados(mLista, oneBeacon)
                mAdapter = AdapterBluetooth(
                    this@BuscandoDispositivosBluetooth,
                    R.layout.adapter_dispositivos_encontrados_wifi,
                    mLista
                )
                listaBluetooth!!.adapter = mAdapter
            }
        }
        try {
            beaconManager!!.startMonitoringBeaconsInRegion(region)
            beaconManager!!.startMonitoringBeaconsInRegion(region2)
        } catch (e: RemoteException) {
            e.printStackTrace()
        }
    }

    // elimina el duplicados de dispositivos encontrados
    fun eliminarDuplicados(
        lista: MutableList<BluetoothFrame>,
        oneBeacon: Beacon
    ): MutableList<BluetoothFrame> {
        val c = Calendar.getInstance()
        @SuppressLint("SimpleDateFormat") val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val strDate = sdf.format(c.time)
        for (i in mLista.indices) {
            if (mLista[i].identifier == oneBeacon.id1) {
                mLista.removeAt(i)
            }
        }

        val frame = BluetoothFrame().apply {
            identifier = oneBeacon.id1
            distance = oneBeacon.distance
            date = strDate
        }

        mLista.add(frame)
        return lista
    }

    // valida los permisos de localizacion
    fun checkPermission() {
        if (Build.VERSION.SDK_INT >= 23) {
            if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED && checkSelfPermission(
                    Manifest.permission.ACCESS_COARSE_LOCATION
                ) == PackageManager.PERMISSION_GRANTED
            ) {
            } else {
                ActivityCompat.requestPermissions(
                    this, arrayOf(
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION
                    ), 1
                )
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED && grantResults[1] == PackageManager.PERMISSION_GRANTED) {
        } else {
            checkPermission()
        }
    }

    @SuppressLint("MissingPermission")
    private fun locationUpdates() {
        val lm = getSystemService(LOCATION_SERVICE) as LocationManager

        try {
            lm.requestLocationUpdates(
                LocationManager.GPS_PROVIDER, 100, 0f
            ) { location ->
                //longitude = location.longitude
                //latitude = location.latitude
                println("La ubicación ha cambiado chaval")
                //println("longitude: $longitude")
                //println("latitude: $latitude")
            }
        } catch (e: Exception) {
            Log.v("Sergio", e.toString())
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        this.googleMap = googleMap
        googleMap.mapType = GoogleMap.MAP_TYPE_NORMAL
        googleMap.setMaxZoomPreference(100f)
        googleMap.setMinZoomPreference(5f)
    }

    companion object {
        private const val TAG = "Sergio"

        //transforma la MAC para que quede Hexadecimal
        fun trasformarMac(): String {
            var mac = ""
            Log.v("Sergio", "uuid|" + (obtenerMac() == "") + "|")
            return if (obtenerMac() == "") {
                alternativaMac()
            } else {
                val macArray = obtenerMac().split(":").toTypedArray()
                for (i in macArray.indices) {
                    if (macArray[i].length == 1) {
                        macArray[i] = macArray[i] + "0"
                    }
                }
                for (i in macArray) {
                    mac = mac + i
                }
                mac
            }
        }

        //TODO revisar traducción
        // optiene la MAC de los dispositivos si es posible
        fun obtenerMac(): String {
            try {
                val all: List<NetworkInterface> =
                    Collections.list(NetworkInterface.getNetworkInterfaces())
                for (nif in all) {
                    if (!nif.name.equals("wlan0", ignoreCase = true)) continue
                    val macBytes = nif.hardwareAddress ?: return ""
                    val res1 = StringBuilder()
                    for (b in macBytes) {
                        res1.append(Integer.toHexString((b and 0xFF.toByte()).toInt()) + ":")
                    }
                    if (res1.length > 0) {
                        res1.deleteCharAt(res1.length - 1)
                    }
                    return res1.toString()
                }
            } catch (ex: Exception) {
                Log.e("Error", ex.message!!)
            }
            return ""
        }

        // concatena los numeros aleatorios
        fun alternativaMac(): String {
            return numeroAleatorio() + numeroAleatorio()
        }

        // se genera numero aletario de 6 digitos
        fun numeroAleatorio(): String {
            val min_val: Long = 100000
            val max_val = 999999
            val randomNum = Math.random() * (max_val - min_val)
            println("Random Number: $randomNum")
            return randomNum.toInt().toString()
        }
    }
}